### ISSUE TITLE
Compose.io. Do we need MONGO_OPLOG_URL? Or just MONGO_URL?

### ISSUE THREAD

**Issue by kafechew** [arunoda/meteor-up-legacy#245]
I was trying Compose.io - Oplog.
Reading back the same issue, same like Compose already change the mongo_url.
So I will get only Replica set URI.

For the meteor up env,
When I put the replica uri as MONGO_URL (ignore MONGO_OPLOG_URL), deployment without issue.

However, when I put replica uri as MONGO_URL and replica uri as MONGO_OPLOG_URL, deployment failed.

What is the right way to setup oplog with compose.io and meteorup?
Thanks!

**Comment by arunoda** [comment]
No you need to put both.
What was the issue?

**Comment by arunoda** [comment]
try to update mup, we did a related fix.

**Comment by kafechew** [comment]
I retried again to put my compose.io replica uri into both MONGO_URL and MONGO_OPLOG_URL but the deployment still failed. Then I removed the MONGO_OPLOG_URL and deploy again, it deployed successfully.

Error message:

```
    [169.53.240.52] ✘ Invoking deployment process: FAILED

-----------------------------------STDERR-----------------------------------
-Dnode_root_dir=/root/.node-gyp/0.10.33',
gyp info spawn args   '-Dmodule_root_dir=/opt/anong/tmp/bundle/programs/server/npm/npm-bcrypt/node_modules/bcrypt',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
gyp info ok 
npm WARN package.json meteor-dev-bundle@0.0.0 No description
npm WARN package.json meteor-dev-bundle@0.0.0 No repository field.
npm WARN package.json meteor-dev-bundle@0.0.0 No README data
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
        0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 80: Connection refused
Latest deployment failed! Reverted back to the previous version.
-----------------------------------STDOUT-----------------------------------
fish.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node: Finished
  COPY Release/bcrypt_lib.node
make: Leaving directory `/opt/anong/tmp/bundle/programs/server/npm/npm-bcrypt/node_modules/bcrypt/build'

> fibers@1.0.1 install /opt/anong/tmp/bundle/programs/server/node_modules/fibers
> node ./build.js

`linux-x64-v8-3.14` exists; testing
Binary is fine; exiting
underscore@1.5.2 node_modules/underscore

semver@2.2.1 node_modules/semver

source-map-support@0.2.5 node_modules/source-map-support
└── source-map@0.1.29 (amdefine@0.1.0)

fibers@1.0.1 node_modules/fibers
Waiting for MongoDB to initialize. (5 minutes)
connected
anong stop/waiting
anong start/running, process 11545
Waiting for 15 seconds while app is booting up
Checking is app booted or not?
anong stop/waiting
anong start/running, process 11630
----------------------------------------------------------------------------
```

Hope to be enlightened. Thanks!

**Comment by arunoda** [comment]
Are you using the latest mup.
If so, you can check logs with _mup logs -n 500_

**Comment by kafechew** [comment]
Yes~ I create a new folder and mup init again.

Error log:

```
 Error: $MONGO_OPLOG_URL must be set to the 'local' database of a Mongo replica set
     at Error (<anonymous>)
     at _.extend._startTailing (packages/mongo/oplog_tailing.js:177)
     at new OplogHandle (packages/mongo/oplog_tailing.js:55)
     at new MongoConnection (packages/mongo/mongo_driver.js:196)
     at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4)
     at Object.<anonymous> (packages/mongo/remote_collection_driver.js:44)
     at Object.defaultRemoteCollectionDriver (packages/underscore/underscore.js:750)
     at new Mongo.Collection (packages/mongo/collection.js:98)
     at packages/accounts-base/accounts_common.js:140
     at /opt/anong/app/programs/server/packages/accounts-base.js:215:4
 error: Forever detected script exited with code: 8
 error: Script restart attempt #49

 /opt/anong/app/programs/server/node_modules/fibers/future.js:173
                        throw(ex);
                          ^
```

The same error keep repeating... till attempt #76

**Comment by arunoda** [comment]
Seems like your mongo oplog is incorrect. It should be point to the local
DB.

**Comment by kafechew** [comment]
I'm using compose.io. I was assigned only one url, which is the replica set uri. How to get the local db url for the MONGO_OPLOG_URL? Thanks!

**Comment by arunoda** [comment]
check this: https://blog.compose.io/meteor-the-oplog-and-elastic-deployment/

**Comment by kafechew** [comment]
This blog is actually outdated... Compose.io had updated the MONGO URL structure. 

What I have now (the only url) is something like mongodb://<user>:<password>@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/anong_oplog?replicaSet=set-jsdjsh64674hvhjjcjd848

When I asked compose.io staff, they said use same url for both MONGO_URL and MONGO_OPLOG_URL will do, but not the case. 

PS.
A note from compose.io:
     The replica set URL is what you're looking for. All Elastic Deployments are in a replica set, so you have two servers running everything in the event one fails. You should configure your application/driver to use the replica set URL as provided in the Admin panel as you've already seen.

**Comment by arunoda** [comment]
Not it's not outdated.
This is the format:

mongodb://<user>:<password>@candidate.11.mongolayer.com:10240,
candidate.0.mongolayer.com:10240/local?authSource=wiktory

This blog is correct.

**Comment by kafechew** [comment]
This and the only url is given by Compose.io:

mongodb://:@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/anong_oplog?replicaSet=set-jsdjsh64674hvhjjcjd848

If that is true...
So, my guess of  MONGO_OPLOG_URL will be like:

mongodb://:@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/local?authSource=anong_oplog?replicaSet=set-jsdjsh64674hvhjjcjd848 ????

I just added "local?authSource" into the default link.
Do you think so? Thanks.

**Comment by arunoda** [comment]
May be you need to convert second ? into &

**Comment by kafechew** [comment]
Hmmm... both version were not working...

mongodb://:@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/local?authSource=anong_oplog?replicaSet=set-jsdjsh64674hvhjjcjd848

mongodb://:@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/local?authSource=anong_oplog&replicaSet=set-jsdjsh64674hvhjjcjd848

Not sure how others use Meteor up + Compose.io new replica set url for Oplog... 
Will keep looking around~

**Comment by arunoda** [comment]
Our replica set URL(using compose) does not have replicaSet param. May be
they have change stuff a bit.
But for Meteor oplog url needs to points to the local database.

**Comment by kafechew** [comment]
Yes. The latest one is with replicaSet param... 
Thanks. Hopefully, can find the solution soon. 
Maybe will be useful for others who face the same issues soon... 
Will keep you updated.

**Comment by arunoda** [comment]
That's great.
Thanks.

**Comment by boustanihani** [comment]
Hey I am also using compose and I use this syntax:

```
"MONGO_URL": "\"mongodb://<username>:<password>@<server>:<port>,<server>:<port>/<database>?replicaSet=<set>&readPreference=primaryPreferred\""

"MONGO_OPLOG_URL": "\"mongodb://<username>:<password>@<server>:<port>,<server>:<port>/local?authSource<database>\""
```

I use **extra double quotes to prevent special characters from being escaped**.
(for example the ampersand "&")

Examples:

```
"MONGO_URL": "\"mongodb://user:123456@candidate.11.mongolayer.com:10240,candidate.0.mongolayer.com:10240/mydatabase?replicaSet=set-0123456789&readPreference=primaryPreferred\""

"MONGO_OPLOG_URL": "\"mongodb://user:123456@candidate.11.mongolayer.com:10240,candidate.0.mongolayer.com:10240/local?authSource=mydatabase\""
```

Check: https://blog.compose.io/meteor-the-oplog-and-elastic-deployment/

**Comment by kafechew** [comment]
@boustanihani Did you make the deployment success with this changes???

I didn't.

```
-----------------------------------STDERR-----------------------------------
 No README data

/usr/lib/node_modules/wait-for-mongo/node_modules/mongodb/lib/mongodb/connection/url_parser.js:15
    throw Error("URL must be in the format mongodb://user:pass@host:port/dbnam
          ^
Error: URL must be in the format mongodb://user:pass@host:port/dbname
    at Error (<anonymous>)
    at exports.parse (/usr/lib/node_modules/wait-for-mongo/node_modules/mongodb/lib/mongodb/connection/url_parser.js:15:11)
    at Function.MongoClient.connect (/usr/lib/node_modules/wait-for-mongo/node_modules/mongodb/lib/mongodb/mongo_client.js:167:16)
    at connectAgain (/usr/lib/node_modules/wait-for-mongo/lib/waitForMongo.js:25:17)
    at waitForMongo (/usr/lib/node_modules/wait-for-mongo/lib/waitForMongo.js:23:3)
    at Object.<anonymous> (/usr/lib/node_modules/wait-for-mongo/bin/wait-for-mongo:12:1)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
-----------------------------------STDOUT-----------------------------------
dule 'bcrypt' of package 'npm-bcrypt'
make: Entering directory `/opt/anong/tmp/bundle/programs/server/npm/npm-bcrypt/node_modules/bcrypt/build'
  CXX(target) Release/obj.target/bcrypt_lib/src/blowfish.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node: Finished
  COPY Release/bcrypt_lib.node
make: Leaving directory `/opt/anong/tmp/bundle/programs/server/npm/npm-bcrypt/node_modules/bcrypt/build'

> fibers@1.0.1 install /opt/anong/tmp/bundle/programs/server/node_modules/fibers
> node ./build.js

`linux-x64-v8-3.14` exists; testing
Binary is fine; exiting
underscore@1.5.2 node_modules/underscore

semver@2.2.1 node_modules/semver

source-map-support@0.2.5 node_modules/source-map-support
└── source-map@0.1.29 (amdefine@0.1.0)

fibers@1.0.1 node_modules/fibers
Waiting for MongoDB to initialize. (5 minutes)
----------------------------------------------------------------------------
```

**Comment by boustanihani** [comment]
@kafechew You're right, the extra double quotes seem to be a problem now, this used to work before, I just tried removing them and deployed again with success:

```
"MONGO_URL": "mongodb://user:123456@candidate.11.mongolayer.com:10240,candidate.0.mongolayer.com:10240/mydatabase?replicaSet=set-0123456789&readPreference=primaryPreferred"

"MONGO_OPLOG_URL": "mongodb://user:123456@candidate.11.mongolayer.com:10240,candidate.0.mongolayer.com:10240/local?authSource=mydatabase"
```

**Comment by kafechew** [comment]
@boustanihani so sad... even using your latest format, I still unable to make it working yet... 

Is there any special setting or node version to make it working?

**Comment by arunoda** [comment]
You need to generate a new user for the oplog access. Check that.

**Comment by kafechew** [comment]
@arunoda I did that (add user w/ oplog access) before I asking this issue. 
So, I still figuring what I had missed @_@

**Comment by arunoda** [comment]
I just deployed an app with new compose replica set for a BulletProof
Meteor lesson. I could do it without any issues.

**Comment by boustanihani** [comment]
@kafechew I am using node version **v0.10.33** on ubuntu 14.04
Check node versions and try running `mup setup` again before deploying...

Are you using the latest meteor ??

**Comment by kafechew** [comment]
Thanks to @arunoda and @boustanihani 
I'm slapping myself now...
I created a new user w/ oplog access and this time work!
Apparently, I didn't add the oplog access to my previous user which I thought I did :-P

So sorry for that and thanks a lot for your patiently helps!!!

Summary: [Meteor with Oplog (Compose.io + MUP)](http://ano.ng/SLRBJ8kRXkqbuer5j)

**Comment by tonyaom** [comment]
For prospective readers and my reference in the future, as at 3/6/2015, the current Connection Strings obtained from compose.io do work. More information can be found in a bit-out-of-dated blog:
https://www.compose.io/articles/meteor-and-compose-a-beginners-guide-to-using-them-together/

Below is my connection strings:
<code>
"MONGO_URL":"mongodb://user:password@c1067.candidate.13.mongolayer.com:11067,candidate.12.mongolayer.com:11102/mydatabasename?replicaSet=set-345ece12345d0d5acd000c3a"

"MONGO_OPLOG_URL":"mongodb://user:password@c1067.candidate.13.mongolayer.com:11067,candidate.12.mongolayer.com:11102/local?authSource=mydatabasename"
</code>
Good luck!

**Comment by jrudio** [comment]
@tonyaom Where did you get `?replicaSet=set-345ece12345d0d5acd000c3a` from? That doesn't show up in the admin section. I'm using a sandbox database, btw.

**Comment by tonyaom** [comment]
@jrudio: it is included in my connection strings on admin section and I am using 30 day free trial

**Comment by derouck** [comment]
I had some issues as well, turned out to be an issue with the compose users:
- Make a separate user for the regular Mongo  without readonly and oplog
- Make a separate user for oplog and make sure the password contains only alphanumeric characters, you must obviously select oplog and you can choose readonly as well.

Everything else is already said often before, hope this helps someone. Good luck.

**Comment by sudiarth** [comment]
mongodb://:@c000.horace.0.mongolayer.com:10138,c000.horace.1.mongolayer.com:10138/anong_oplog?replicaSet=set-jsdjsh64674hvhjjcjd848

Its work by add this repeated line to mup.json

"MONGO_URL": "mongodb://user:password@c000.horace.0.mongolayer.com:10138/anong_oplog",
"MONGO_URL": "mongodb://user:password@c000.horace.1.mongolayer.com:10138/anong_oplog",

**Comment by kahmali** [comment]
@jrudio I'm sure you've figured this out by now, but Mongo doesn't use replica sets for the sandbox deployments. You have to use a regular deployment to get access to replica sets. It will be created for you automatically with the regular MongoDB deployment, if I remember correctly. Just mentioning this in case it helps anyone in the future.

**Comment by derwaldgeist** [comment]
Is there any way to tell whether Meteor is actually _using_ the oplog?

**Comment by kahmali** [comment]
@derwaldgeist I'm not sure if this is what you're looking for, but if you check the Live Queries tab in your Kadira Dashboard, you can sort the publications by the total number of oplog notifications, to see where the oplog is most active. There is also a graph (the third one down in the image below) showing the oplog notification count over a designated interval. 
<img width="1342" alt="screenshot 2016-01-15 13 58 14" src="https://cloud.githubusercontent.com/assets/935297/12362375/f63953f8-bb90-11e5-8ac3-fa26d56c2dd2.png">

You can also see the oplog being used by exploring certain method traces, in the Methods tab of your Kadira dashboard:
<img width="588" alt="screenshot 2016-01-15 14 04 40" src="https://cloud.githubusercontent.com/assets/935297/12362415/325358c0-bb91-11e5-9041-679546c9bca4.png">

**Comment by MasterJames** [comment]
There's this link you might learn some more from.
https://meteorhacks.com/mongodb-oplog-and-meteor/
I recall you have to enable it but I didn't thing meteor used even a 64bit version by default yet so I thought you have to setup mongo more manually and use something like mupx to point your deployment to it. 
Also note the meteorhacks "cluster" plugin. I've not fully assesses all this yet but I believe the oplog is key especially with a shared filesystem like AWS' EFS which will be glorious... one day hopefully soon but without EFS for everyone (general access?) there will be limited demand push and progress on oplog and the best solution for now is the replica set, which means a master with failure recovery when it fails. It just doesn't sound like a good load handling solution with a main master getting all DB traffic (if I'm understanding properly yet). So something like cluster makes good sense. 
I recall seeing or putting a setting to disable the oplog (which suggest it's on by default) I think in my PR for mongodb backup and restore functionality because it was causing problems with restore if it's on or something like that.
Basically you need to make a replica set and that's not what mupx does I don't think.

Google for more here's something. ..
https://themeteorchef.com/snippets/setting-up-mongodb-oplog-tailing/
And
http://www.manuel-schoebel.com/blog/meteorjs-and-mongodb-replica-set-for-oplog-tailing

**Comment by derwaldgeist** [comment]
@kahmali @MasterJames Thanks!
